velero-plugins/pod/restore.go

```diff
@@ -29,6 +31,64 @@ func (p *RestorePlugin) AppliesTo() (velero.ResourceSelector, error) {
 	}, nil
 }
 
+// Returns true if pod is not excluded by namespace rules
+func (p *RestorePlugin) podRestoreHookIncludeNamespace(pod *corev1API.Pod, restoreHookSpec velerov1.RestoreResourceHookSpec) bool {
+	if len(restoreHookSpec.IncludedNamespaces) == 0 && !common.StringInSlice(pod.Namespace, restoreHookSpec.ExcludedNamespaces) {
+		p.Log.Info("[pod-restore] pod has restore hooks via unspecified included namespaces and is not in excluded namespaces")
+		return true
+	}
+	if common.StringInSlice(pod.Namespace, restoreHookSpec.IncludedNamespaces) {
+		p.Log.Info("[pod-restore] pod has restore hooks via included namespaces")
+		return true
+	}
+	p.Log.Info("[pod-restore] pod has no restore hooks via included/exclude namespaces")
+	return false
+}
+
+func (p *RestorePlugin) podRestoreHookIncludeResources(pod *corev1API.Pod, restoreHookSpec velerov1.RestoreResourceHookSpec) bool {
+	if len(restoreHookSpec.IncludedResources) == 0 && !common.StringInSlice(pod.Name, restoreHookSpec.ExcludedResources) {
+		p.Log.Info("[pod-restore] pod has restore hooks via unspecified included resources and is not in excluded resources")
+		return true
+	}
+	if common.StringInSlice(pod.Name, restoreHookSpec.IncludedResources) {
+		p.Log.Info("[pod-restore] pod has restore hooks via included resources")
+		return true
+	}
+	p.Log.Info("[pod-restore] pod has no restore hooks via included/exclude resources")
+	return false
+}
+
+// Check if pod has restore hooks via pod annotations or via restore hook rules
+func (p *RestorePlugin) podHasRestoreHooks(pod corev1API.Pod, resources []velerov1.RestoreResourceHookSpec) (bool, error) {
+	_, postRestoreHookDefined := pod.Annotations[common.PostRestoreHookAnnotation]
+	_, initContainerRestoreHookDefined := pod.Annotations[common.InitContainerRestoreHookAnnotation]
+	if postRestoreHookDefined || initContainerRestoreHookDefined {
+		p.Log.Info("[pod-restore] pod has restore hooks via annotations")
+		return true, nil
+	}
+	p.Log.Info("[pod-restore] pod has no restore hooks via annotations")
+	
+	for _, restoreHookSpec := range resources {
+		p.Log.Debugf("DEBUG [pod-restore] %v", restoreHookSpec)
+		if  len(restoreHookSpec.PostHooks) == 0 {
+			continue
+		}
+		//convert MatchLabels to labels.Selector
+		selector, err := metav1.LabelSelectorAsSelector(restoreHookSpec.LabelSelector)
+		if err != nil {
+			p.Log.Errorf("[pod-restore] labelselector conversion error: %v", err)
+			return false, err
+		}
+		if  len(restoreHookSpec.PostHooks) > 0 && selector.Matches(labels.Set(pod.Labels)) ||
+			(p.podRestoreHookIncludeNamespace(&pod, restoreHookSpec) &&	p.podRestoreHookIncludeResources(&pod, restoreHookSpec)) {
+			p.Log.Info("[pod-restore] pod has restore hooks via include/exclude rules")
+			return true, nil
+		}
+	}
+	p.Log.Info("[pod-restore] pod has no restore hooks")
+	return false, nil
```

Comment:
Looks like you're treating the resource spec conditions as an OR rather than an AND. If you look at the velero item_hook_handler (which actually decides whether to call the hook), it's an and. The namespace, resource, and label selector must all match. Basically the if conditions need to be reversed, returning false on failure, and return true at the end if they all succeeded.

Reply:
+1